### DISCUSSION TITLE
Can kubernetes-client work with random resources or be extended to do so?

### DISCUSSION THREAD

**Discussion by alekseiAlefirov** [joan38/kubernetes-client#176] [Q&A]
Hi Joan, team,

First of all, thanks for your great work and this awesome library! Having typelevel-stack tool to access Kubernetes API with (among other features) an option to configure authorization credentials is very handy.

But I have found it a bit problematic to use for anything beyond common entities APIs (pods, services, etc.).
So my task was to access Openshift [_routes_](https://docs.openshift.com/container-platform/3.11/rest_api/route_openshift_io/route-route-openshift-io-v1.html) from a k8s cluster. I tried it with `CustomResourcesApi`, using circe Json type for the insides, as it was easy for quick testing. But it broke on parsing with
```
DecodingFailure(Attempt to decode value on failed cursor, List(DownField(apiVersion), DownArray, DownField(items)))
```
As you can see from [_routes_](https://docs.openshift.com/container-platform/3.11/rest_api/route_openshift_io/route-route-openshift-io-v1.html) specification, an item of them has only one mandatory field, `spec`, while `apiVersion` and `kind`, required by `CustomResource`, are  optional. Which caused the problem for me -- they were missing in the fetched data, and client failed to parse it.

I thought then, OK, I can try to extend the API, make another `Listable`. But I was sad to see, that `Listable` is private! And so are all `-able` traits, which appears to make impossible to use `kubernetes-client` to work with more or less non-canon data.

Can these traits become public? Is it possible to extend `kubernetes-client` to digest more random data (keeping structure with `-able` traits)? Or am I missing some point about the library?

**Comment by joan38** [comment]
Hey, I think this is supported now by #16 from @novakov-alexey-zz
Let me know if you have any question. Closing in the meantime.
